apps/settings/js/panels/improve_browser_os/panel.js

```diff
@@ -1,37 +1,97 @@
+          settings.createLock().get(METRICLEVELSETTING);
+        if (metricsLevelReq) {
+          metricsLevelReq.onsuccess = function() {
+            var result = metricsLevelReq.result[METRICLEVELSETTING];
+            if (result === 'Basic') {
+              metricsBasic.checked = true;
+            } else if (result === 'Enhanced') {
+              metricsEnhanced.checked = true;
+            } else if (result === 'None') {
+              metricsNone.checked = true;
+            } else {
+              // When upgrading from a release that has only the
+              // 'share data' checkbox, use the 'sharing' status
+              //  to determine the default metric level.
+              var sharingReq = settings &&
+                settings.createLock().get(SHAREDATASETTING);
+              if (sharingReq) {
+                sharingReq.onsuccess = function() {
+                  var result = sharingReq.result[SHAREDATASETTING];
+                  if (result) {
+                    metricsBasic.checked = true;
+                  } else {
+                    metricsNone.checked = true;
+                  }
+                };
+              }
+            }
+          };
+        }
+
+        // Dogfooders should not be able to disable the metrics.
+        var dogfood = settings && settings.createLock().get(DOGFOODSETTING);
+        if (dogfood) {
+          dogfood.onsuccess = function() {
+            if (dogfood.result[DOGFOODSETTING]) {
+              metricsBasic.setAttribute('disabled', 'true');
+              metricsEnhanced.setAttribute('disabled', 'true');
+              metricsNone.setAttribute('disabled', 'true');
+            } else {
+              metricsBasic.removeAttribute('disabled');
+              metricsEnhanced.removeAttribute('disabled');
+              metricsNone.removeAttribute('disabled');
+            }
+          };
+        }
+
       },
 
-      _openAppUsage: function() {
-        var activity = new window.MozActivity({
-          name: 'marketplace-search',
-          data: {
-            type: 'firefox-os-app-stats'
-          }
-        });
-        activity.onerror = function() {
-          console.log('Error opening app stats page in Marketplace');
-        };
+      handleEvent: function (event) {
+        switch (event.target.id) {
+          case 'appUsageByRegion':
+            var activity = new window.MozActivity({
+              name: 'marketplace-search',
+              data: {
+                type: 'firefox-os-app-stats'
+              }
+            });
+            activity.onerror = function() {
+              console.log('Error opening app stats page in Marketplace');
+            };
+            break;
+
+          case 'metrics-basic':
+          case 'metrics-enhanced':
+          case 'metrics-none':
+            var metricLevel = {};
+            metricLevel[METRICLEVELSETTING] = event.target.value;
+            SettingsListener.getSettingsLock().set(metricLevel);
```

Comment:
with `settings.createLock().set`, you don't have to introduce `SettingsListener`